File: Backend/app/models/order.py
from typing import Annotated, Any, List, Optional, Dict
from pydantic import BaseModel, Field, BeforeValidator
from bson import ObjectId

def validate_object_id(v: Any) -> str:
    if isinstance(v, ObjectId):
        return str(v)
    if isinstance(v, str) and ObjectId.is_valid(v):
        return v
    raise ValueError("Invalid ObjectId")

PyObjectId = Annotated[str, BeforeValidator(validate_object_id)]

class Address(BaseModel):
    address: str
    city: str
    country: str

class OrderItem(BaseModel):
    product: PyObjectId  # Changed from product_id to match service class
    quantity: int
    price: float

class Order(BaseModel):
    id: PyObjectId = Field(default_factory=lambda: str(ObjectId()), alias="_id")
    user_id: PyObjectId = Field() 
    items: List[OrderItem]
    total_price: float
    status: str  # pending, completed, cancelled
    payment_method: str  # credit_card, paypal, bank_transfer
    payment_status: str  # pending, completed, failed
    shipping_address: Address

    model_config = {
        "populate_by_name": True,
        "json_schema_extra": {
            "example": {
                "_id": "67c16c4ed87c5f25d3ae561c",
                "user_id": "67be899eabfe0be00df5b9dd",
                "items": [
                    {
                        "product": "67b3011e64ce74113d1500a8",  # Changed from product_id
                        "quantity": 2,
                        "price": 8.99
                    }
                ],
                "total_price": 17.98,
                "payment_method": "paypal",
                "payment_status": "pending",
                "shipping_address": {
                    "address": "123 Main St",
                    "city": "Hanoi",
                    "country": "Vietnam",
                },
            }
        }
    }

File: ecommercial_shopping/lib/core/models/order.dart
class Order {
  final String id;
  final Address shippingAddress;
  final List<OrderItem> items;
  final String paymentMethod;
  final String paymentStatus;
  final double totalPrice;
  final String userId;

  Order({
    required this.id,
    required this.shippingAddress,
    required this.items,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.totalPrice,
    required this.userId,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    var itemsJson = json['items'] as List? ?? [];
    List<OrderItem> itemsList =
        itemsJson.map((item) => OrderItem.fromJson(item)).toList();

    return Order(
      id: json['_id'] ?? '',
      shippingAddress: Address.fromJson(json['shipping_address'] ?? {}),
      items: itemsList,
      paymentMethod: json['payment_method'] ?? '',
      paymentStatus: json['payment_status'] ?? '',
      totalPrice: (json['total_price'] ?? 0).toDouble(),
      userId: json['user_id'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'shipping_address': shippingAddress.toJson(),
      'items': items.map((item) => item.toJson()).toList(),
      'payment_method': paymentMethod,
      'payment_status': paymentStatus,
      'total_price': totalPrice,
      'user_id': userId,
    };
  }
}

class Address {
  final String address;
  final String city;
  final String country;

  Address({
    required this.address,
    required this.city,
    required this.country,
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    return Address(
      address: json['address'] ?? '',
      city: json['city'] ?? '',
      country: json['country'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'address': address,
      'city': city,
      'country': country,
    };
  }
}

class OrderItem {
  final String productId;
  final int quantity;
  final double price;

  OrderItem({
    required this.productId,
    required this.quantity,
    required this.price,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      productId: json['product'] ?? '',
      quantity: json['quantity'] ?? 0,
      price: (json['price'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'product': productId,
      'quantity': quantity,
      'price': price,
    };
  }
}
